pkg/datapath/linux/arp/ping.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018-2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arp
+
+import (
+	"bytes"
+	"errors"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"net"
+	"time"
+)
+
+var (
+	ErrNotImplemented = errors.New("not implemented")
+	ErrL2Unreachable  = errors.New("interface can't reach the IP address")
+
+	timeout = 1 * time.Second
+)
+
+var defaultSerializeOpts = gopacket.SerializeOptions{
+	FixLengths:       true,
+	ComputeChecksums: true,
+}
+
+// PingOverIface performs arping request to the 'dst' IP address over the 'iface' and
+// returns the hardware address (MAC) of the destination if it is reachable
+func PingOverIface(iface net.Interface, dst net.IP) (net.HardwareAddr, error) {
+	src, err := findSourceIP(iface, dst)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := newPinger(&iface, src)
+	if err != nil {
+		return nil, err
+	}
+	defer p.close()
+
+	if err := p.setDeadline(time.Now().Add(timeout)); err != nil {
+		return nil, err
+	}
+
+	return p.resolve(dst)
+}
+
+var _ net.Addr = &Addr{}
+
+type Addr struct {
+	net.HardwareAddr
+}
+
+func (a *Addr) Network() string {
+	return "raw"
+}
+
+type pinger struct {
+	c     net.PacketConn
+	ip    net.IP
+	iface *net.Interface
```

Comment:
Could we use `*netlink.Link` instead? Asking, as in `pkg/datapath/linux/node.go` we do `netlink.LinkByName()` before the arping, so requiring to call `net.InterfaceByName()` seems to be a bit wasteful.

Reply:
I think we could, but the problem is that that `netlink.LinkByName()` call is done between two calls to the `arping` package (one to check if an IP is L2 reachable and another to do the ping). So, I decided to combine those two calls as logically the check for the reachability, could be a part of the  `arp.resolve()` method. If it is decided that `node.go` should be left untouched, I will still need to have a method which checks the reachability in the new `arp` package. So the package will be tightly coupled to a method from the `node.go`. Please let me know if using the `*netlink.Link` would still be preferable.

Reply:
OK, I meant that `PingOverIface(...)` should accept `*netlink.Link` instead of `net.Interface`. This should completely eliminate `net.Interface` in the package.

Reply:
> OK, I meant that PingOverIface(...) should accept *netlink.Link instead of net.Interface. This should completely eliminate net.Interface in the package.

I understood and don't have a strong preference. However, after additional thought, if `*netlink.Link` is used as a parameter of `PingOverIface()`, even if that compiles on other platform than Linux, the communicated intention may appear as Linux specific. I'm not sure what are the plans of the project on building/supporting something else than Linux, but I just wanted to mention that.

It appears that if I go the suggested way, I would need to make `findSourceIP()` a part of the `node.go` and PingOverIface will accept a source IP in addition to the `*netlink.Link`.  How does it sound to you?

Reply:
> However, after additional thought, if *netlink.Link is used as a parameter of PingOverIface(), even if that compiles on other platform than Linux

I think it's not a problem. We are already extensively using `*netlink.Link` in other parts of Cilium.

> I would need to make findSourceIP() a part of the node.go and PingOverIface will accept a source IP in addition to the *netlink.Link

Why cannot it stay in the same module?

Reply:
> Why cannot it stay in the same module?

The module would not depend on `findSourceIP()` if `PingOverIface()` receives the source IP as a parameter and, in my opinion, it would be strange to find a function unrelated to ARP operations in the module called `arp`. However, I can make that change if you confirm that, it will be acceptable.

Reply:
My reasoning was that finding a source IP addr is one more additional thing which a user of the package would need to take care of. But this is not the case for us, as we should already know the source IP from the route: https://github.com/cilium/cilium/blob/master/pkg/datapath/linux/node.go#L568. So, OK, let's make the function to accept the source IP as a param.

Reply:
@brb, I've made the changes as you requested. Let me know if something else needs an update.